File: swift/FeedbackSheet.swift
import SwiftUI

// MARK: - FeedbackSheet
// A bottom sheet with a feedback text field, optional email field,
// send button with cooldown, and character counter.

struct FeedbackSheet: View {
    @Environment(\.dismiss) private var dismiss

    // Configurable strings
    var feedbackPlaceholder: String = "What's on your mind?"
    var emailPlaceholder: String = "Leave your email if you'd like us to follow up (totally optional)"
    var sendButtonText: String = "Send"
    var successMessage: String = "Thank you for your feedback!"
    var errorMessage: String = "Could not send feedback. Please check your connection and try again."

    @State private var feedbackText = ""
    @State private var emailText = ""
    @State private var isSending = false
    @State private var showSuccess = false
    @State private var showError = false
    @State private var cooldownRemaining = 0
    @State private var cooldownTimer: Timer?

    @FocusState private var focusedField: Field?

    private let maxCharacters = 2000
    private let cooldownDuration = 60
    private let cooldownKey = "feedpush_last_send_timestamp"

    private enum Field {
        case feedback, email
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(spacing: 20) {
                    feedbackField
                    emailField
                    sendButton
                }
                .padding(20)
            }
            .scrollDismissesKeyboard(.interactively)
            .navigationTitle("Feedback")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Close") { dismiss() }
                }
            }
            .overlay {
                if showSuccess {
                    successOverlay
                }
            }
            .alert("Error", isPresented: $showError) {
                Button("OK", role: .cancel) {}
            } message: {
                Text(errorMessage)
            }
            .onAppear { startCooldownIfNeeded() }
            .onDisappear { cooldownTimer?.invalidate() }
            .contentShape(Rectangle())
            .onTapGesture { focusedField = nil }
        }
    }

    // MARK: - Feedback Text Field

    private var feedbackField: some View {
        VStack(alignment: .leading, spacing: 8) {
            TextEditor(text: $feedbackText)
                .focused($focusedField, equals: .feedback)
                .frame(minHeight: 120)
                .padding(12)
                .scrollContentBackground(.hidden)
                .background(Color(.secondarySystemBackground))
                .clipShape(RoundedRectangle(cornerRadius: 12))
                .onChange(of: feedbackText) { _, newValue in
                    if newValue.count > maxCharacters {
                        feedbackText = String(newValue.prefix(maxCharacters))
                    }
                }
                .overlay(alignment: .topLeading) {
                    if feedbackText.isEmpty {
                        Text(feedbackPlaceholder)
                            .foregroundStyle(.tertiary)
                            .padding(.horizontal, 17)
                            .padding(.vertical, 20)
                            .allowsHitTesting(false)
                    }
                }

            HStack {
                Spacer()
                Text("\(feedbackText.count) / \(maxCharacters)")
                    .font(.caption)
                    .foregroundStyle(feedbackText.count > maxCharacters ? .red : .secondary)
            }
        }
    }

    // MARK: - Email Field

    private var emailField: some View {
        TextField(emailPlaceholder, text: $emailText)
            .focused($focusedField, equals: .email)
            .keyboardType(.emailAddress)
            .textContentType(.emailAddress)
            .autocorrectionDisabled()
            .textInputAutocapitalization(.never)
            .padding(14)
            .background(Color(.secondarySystemBackground))
            .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    // MARK: - Send Button

    private var sendButton: some View {
        Button {
            Task { await sendFeedback() }
        } label: {
            Group {
                if isSending {
                    ProgressView()
                        .tint(.white)
                } else if cooldownRemaining > 0 {
                    Text("\(sendButtonText) (\(cooldownRemaining)s)")
                } else {
                    Text(sendButtonText)
                }
            }
            .font(.headline)
            .foregroundStyle(.white)
            .frame(maxWidth: .infinity)
            .padding(16)
            .background(isSendDisabled ? Color.gray : Color.accentColor)
            .clipShape(RoundedRectangle(cornerRadius: 14))
        }
        .disabled(isSendDisabled)
    }

    private var isSendDisabled: Bool {
        feedbackText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        || feedbackText.count > maxCharacters
        || isSending
        || cooldownRemaining > 0
    }

    // MARK: - Success Overlay

    private var successOverlay: some View {
        VStack(spacing: 12) {
            Image(systemName: "checkmark.circle.fill")
                .font(.system(size: 48))
                .foregroundStyle(.green)
            Text(successMessage)
                .font(.headline)
                .multilineTextAlignment(.center)
        }
        .padding(32)
        .background(.ultraThickMaterial)
        .clipShape(RoundedRectangle(cornerRadius: 20))
        .transition(.scale.combined(with: .opacity))
        .padding()
    }

    // MARK: - Send Logic

    private func sendFeedback() async {
        focusedField = nil
        isSending = true

        let email = emailText.trimmingCharacters(in: .whitespacesAndNewlines)

        do {
            try await FeedbackService.send(
                text: feedbackText,
                email: email.isEmpty ? nil : email
            )
            triggerHaptic()
            withAnimation { showSuccess = true }
            saveCooldownTimestamp()
            startCooldownTimer()
            feedbackText = ""
            emailText = ""

            try? await Task.sleep(for: .seconds(2))
            dismiss()
        } catch {
            showError = true
        }

        isSending = false
    }

    // MARK: - Haptics

    private func triggerHaptic() {
        let generator = UINotificationFeedbackGenerator()
        generator.notificationOccurred(.success)
    }

    // MARK: - Cooldown

    private func saveCooldownTimestamp() {
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: cooldownKey)
    }

    private func startCooldownIfNeeded() {
        let lastSend = UserDefaults.standard.double(forKey: cooldownKey)
        guard lastSend > 0 else { return }

        let elapsed = Int(Date().timeIntervalSince1970 - lastSend)
        let remaining = cooldownDuration - elapsed

        if remaining > 0 {
            cooldownRemaining = remaining
            startCooldownTimer()
        }
    }

    private func startCooldownTimer() {
        cooldownTimer?.invalidate()
        cooldownTimer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { _ in
            if cooldownRemaining > 0 {
                cooldownRemaining -= 1
            } else {
                cooldownTimer?.invalidate()
            }
        }
    }
}

#Preview {
    FeedbackSheet()
}

File: kotlin/FeedbackSheet.kt
package com.feedpush

import android.view.HapticFeedbackConstants
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

// MARK: FeedbackSheet
// A modal bottom sheet with a feedback text field,
// send button with cooldown, and character counter.

private const val MAX_CHARACTERS = 2000
private const val COOLDOWN_DURATION = 60
private const val COOLDOWN_KEY = "feedpush_last_send_timestamp"

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FeedbackSheet(
    onDismiss: () -> Unit,
    feedbackPlaceholder: String = "What's on your mind?",
    sendButtonText: String = "Send",
    successMessage: String = "Thank you for your feedback!",
    errorMessage: String = "Could not send feedback. Please check your connection and try again."
) {
    val context = LocalContext.current
    val view = LocalView.current
    val scope = rememberCoroutineScope()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    var feedbackText by remember { mutableStateOf("") }
    var isSending by remember { mutableStateOf(false) }
    var showSuccess by remember { mutableStateOf(false) }
    var showError by remember { mutableStateOf(false) }
    var cooldownRemaining by remember { mutableIntStateOf(0) }

    val prefs = remember { context.getSharedPreferences("feedpush", android.content.Context.MODE_PRIVATE) }

    val isSendDisabled = feedbackText.trim().isEmpty()
            || feedbackText.length > MAX_CHARACTERS
            || isSending
            || cooldownRemaining > 0

    // Check existing cooldown on appear
    LaunchedEffect(Unit) {
        val lastSend = prefs.getLong(COOLDOWN_KEY, 0L)
        if (lastSend > 0) {
            val elapsed = ((System.currentTimeMillis() - lastSend) / 1000).toInt()
            val remaining = COOLDOWN_DURATION - elapsed
            if (remaining > 0) {
                cooldownRemaining = remaining
            }
        }
    }

    // Cooldown timer
    LaunchedEffect(cooldownRemaining) {
        if (cooldownRemaining > 0) {
            delay(1000)
            cooldownRemaining -= 1
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState
    ) {
        Box(modifier = Modifier.fillMaxWidth()) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = 20.dp)
                    .padding(bottom = 32.dp)
            ) {
                // Header
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = "Feedback",
                        fontSize = 20.sp,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.weight(1f)
                    )
                    IconButton(onClick = onDismiss) {
                        Icon(
                            imageVector = Icons.Default.Close,
                            contentDescription = "Close"
                        )
                    }
                }

                Spacer(modifier = Modifier.height(16.dp))

                // Feedback text field
                OutlinedTextField(
                    value = feedbackText,
                    onValueChange = { if (it.length <= MAX_CHARACTERS) feedbackText = it },
                    placeholder = { Text(feedbackPlaceholder) },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(150.dp),
                    shape = RoundedCornerShape(12.dp)
                )

                // Character counter
                Text(
                    text = "${feedbackText.length} / $MAX_CHARACTERS",
                    fontSize = 12.sp,
                    color = if (feedbackText.length > MAX_CHARACTERS) {
                        MaterialTheme.colorScheme.error
                    } else {
                        MaterialTheme.colorScheme.onSurfaceVariant
                    },
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = 4.dp),
                    textAlign = TextAlign.End
                )

                Spacer(modifier = Modifier.height(24.dp))

                // Send button
                Button(
                    onClick = {
                        scope.launch {
                            isSending = true
                            showError = false

                            val result = FeedbackService.send(context, feedbackText)

                            when (result) {
                                is FeedbackResult.Success -> {
                                    view.performHapticFeedback(HapticFeedbackConstants.CONFIRM)
                                    showSuccess = true
                                    prefs.edit().putLong(COOLDOWN_KEY, System.currentTimeMillis()).apply()
                                    cooldownRemaining = COOLDOWN_DURATION
                                    feedbackText = ""
                                    delay(2000)
                                    onDismiss()
                                }
                                is FeedbackResult.Error -> {
                                    showError = true
                                }
                            }

                            isSending = false
                        }
                    },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(52.dp),
                    enabled = !isSendDisabled,
                    shape = RoundedCornerShape(14.dp)
                ) {
                    if (isSending) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(20.dp),
                            strokeWidth = 2.dp,
                            color = MaterialTheme.colorScheme.onPrimary
                        )
                    } else if (cooldownRemaining > 0) {
                        Text("$sendButtonText (${cooldownRemaining}s)")
                    } else {
                        Text(sendButtonText)
                    }
                }

                // Error message
                if (showError) {
                    Spacer(modifier = Modifier.height(12.dp))
                    Text(
                        text = errorMessage,
                        color = MaterialTheme.colorScheme.error,
                        fontSize = 13.sp,
                        modifier = Modifier.fillMaxWidth(),
                        textAlign = TextAlign.Center
                    )
                }
            }

            // Success overlay
            AnimatedVisibility(
                visible = showSuccess,
                enter = fadeIn() + scaleIn(),
                exit = fadeOut() + scaleOut(),
                modifier = Modifier.align(Alignment.Center)
            ) {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    modifier = Modifier.padding(32.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.CheckCircle,
                        contentDescription = "Success",
                        tint = MaterialTheme.colorScheme.primary,
                        modifier = Modifier.size(48.dp)
                    )
                    Spacer(modifier = Modifier.height(12.dp))
                    Text(
                        text = successMessage,
                        fontWeight = FontWeight.Bold,
                        textAlign = TextAlign.Center
                    )
                }
            }
        }
    }
}
